### ISSUE TITLE
column names are not shown if the slickgrid table is not on the screen

### ISSUE THREAD

**Issue by UtmostCreator** [ghiscoding/aurelia-slickgrid#528]
### Your Environment

| Software          | Version(s) |
| ----------------- | ---------- |
| Aurelia           |  2.0.2
| Aurelia-Slickgrid | 3.2.1
| TypeScript        | vanilla JS
| Operating System  | windows 10 20h2
| NPM/Node/Yarn     | 14.15.4

### Context

I have several tabs on in the web-application.
one of my slickgrid tables is there.
If I open this tab, everything goes well except the column names, they are all not displayed until I do next:
1. resize the window.
2.  run this code when the tab is opened:
```
let cols = this.grid.getColumns();
this.grid.setColumns(cols);
```
issue image:
![image](https://user-images.githubusercontent.com/22347676/109209251-2a967000-7760-11eb-835e-7b9eb7350cf5.png)

### Expected Behavior

After I open any of the tabs (where there is a slickgrid table), the slickgrid table's column names is not sized (shown) correctly and do not displayed on a page

### Current Behavior

After I open any of the tabs (where there is a slickgrid table), column names must be sized (shown) correctly on a page

### Possible Solution

But this is just a workaround, I believe there is better solution there.
```
let cols = this.grid.getColumns();
this.grid.setColumns(cols);
```

**Comment by ghiscoding** [comment]
So there's couple things to mention
1. you need to know is that SlickGrid does not work well if you create your grid in a hidden div because it uses that to do the column sizing and everything (I think you can use `opacity: 0` though). 
2. another thing to note (because of number 1.) is that if you have grids in multiple tabs, you have to call a grid resize after changing tab and funny enough, I did that exact same fix in my demo very recently and you can see the fix in this [PR](https://github.com/ghiscoding/aurelia-slickgrid/pull/525) (without the fix, the title weren't showing on the 2nd tab which I think is your problem), you simply need to add a handler when changing tab and call a grid resize (it looks like I already pushed the fix already on the GitHub demo, so you can see it working on [Example 22](https://ghiscoding.github.io/aurelia-slickgrid/#/slickgrid/example22)). 
3.  lastly do not call `grid.setColumns(cols);` directly from SlickGrid because if you do then you are completely bypassing Aurelia-Slickgrid internal code and the column references that Aurelia-Slickgrid keeps internally will be out of sync (I have a ref for `allColumns` and `allVisibleColumns` in my lib) and you might have some weird behaviors here and there if you bypass it the way you did. I mean you could call it but you would be responsible to have your columns in sync. So if you need to modify the column definitions array, then you should follow the [Example 3](https://ghiscoding.github.io/aurelia-slickgrid/#/slickgrid/example3) from this [line](https://github.com/ghiscoding/aurelia-slickgrid/blob/master/src/examples/slickgrid/example3.ts#L588)
![image](https://user-images.githubusercontent.com/643976/109237418-7b798900-779f-11eb-9705-269e0c84cb5b.png)
    - basically it's very good to know what methods to use from SlickGrid but there are couple of exceptions you should be aware of and `setColumns()` is one of those, feel free to use `getColumns()` though

**Comment by UtmostCreator** [comment]
Thank you, Ghislain B.

Sorry if it is off-topic (let me know if I must create a separate issue on this, or we can keep it here). Please could you give advice or a link to a similar situation?
How can I get the last id of the row that was selected/deselected?
I want it to add/remove items from my custom array by selecting/deselecting items from the slickgrid table.

2. Also, is there a way to avoid comparing (selected items see the code down below)
Precondition:
I have a table with items (let's assume we have 4 items selected by default);
Before this table a have a checkbox (show selected/all items)
Desired behavior:
It must show only selected items if checked, and if it is unchecked, all items must be shown (selected and not selected).


```
  selectAssignedDevices() {
    let that = this;
    setTimeout(() => {
      let activeItems = that.dataset.filter(x => {
        let elementsOfArray2PresentInArray1 = that.assignedDevices.filter(y => {
          return y['@rid'] === x.rid;
        });

        if (elementsOfArray2PresentInArray1.length > 0) {
          return true;
        }
        return false;

        //`return !length;` will  return false if length > 0
      });
      activeItems = activeItems.map(function(item) { return item.id; });
      that.grid.setSelectedRows(activeItems);
    }, 0);
  }
```

3. if there is a way to add selected items in a better way?
4. I am getting all the devices asynchronous, and even if I add items like this `that.grid.setSelectedRows(activeItems);` it does not work w/o setTimeout, but this function is called the last.

**Comment by ghiscoding** [comment]
You should ask these kind of questions on Stack Overflow where many more users could answer but to answer some of the questions
- Read [Grid & DataView Events](https://github.com/ghiscoding/aurelia-slickgrid/wiki/Grid-&-DataView-Events#1-example-with-delegate-event-dispatch-asgonx)
   - use the `onSelectedRowsChanged` to get info of what changed (before/after), if you want to know if everything is selected then check its new length against your dataset length I guess that would work without looping through them all
- `setSelectedRows` is the correct way to change selection and if you need to add a `setTimeout` then it's because you have a racing condition (something else is changing or rendering the grid)

**Comment by UtmostCreator** [comment]
Good day!

You have `previousSelectedRows` prop and `rows` on event called `onGridSelectedRowsChanged`.
But It would be nice to have the last selected Row as a single item there.
For now, `rows` contain ids of selected items, and `previousSelectedRows` all previousSelectedRows.
It would be much nicer to see the lastly selected row id (separately or a row itself)

![image](https://user-images.githubusercontent.com/22347676/109666129-bffc8000-7b23-11eb-8257-50a3e6cced82.png)

your Example 22 works just perfect!


Thank you!

**Comment by ghiscoding** [comment]
The `onSelectedRowsChanged` event is not from my lib but from SlickGrid itself (core) lib itself, on this [line](https://github.com/6pac/SlickGrid/blob/master/slick.grid.js#L2799), so if you want some changes then you can contribute and create a PR (Pull Request) on [6pac/SlickGrid](https://github.com/6pac/SlickGrid) fork, the Example 22 was modified with that missing code for that reason.

So I think I have answered all your questions and again if you have more questions, I suggest you ask on Stack Overflow since there a lot more people that can answer (I answer some there too)

Cheers ⭐
